### ISSUE TITLE
Enabling compass sprite classes to be output as placeholders

### ISSUE THREAD

**Issue by superbland** [Compass/compass#1005]
Currently this code:

```
@import "site/*.png";
@include all-site-sprites;
```

Produces an output that contains a list of classes relating to the background image and positions that can be extended. Since the recent addition of placeholders and the move to prevent extending from within media directives, it will be impossible to extend these classes without potentially having massive amounts of duplicate code within the media query.

Using placeholders instead of classes (or allowing a choice) would mean that a mixin containing sprite css could be included without unnecessary duplication.

**Comment by silvenon** [comment]
Unfortunately, Sass doesn't seem to be able to extend placeholders beyond the `@media` block (and probably classes too, I haven't tried it yet). Nevertheless, it's a great suggestion, I'll implement it when I find the time.

Yeah, there will be a choice, the default will be `class` for backwards compatibility.

**Comment by aaronjensen** [comment]
+1

**Comment by silvenon** [comment]
I forgot about this... Don't count on me because I don't use Compass anymore.

**Comment by scottdavis** [comment]
Well that is sad news

**Comment by silvenon** [comment]
Probably not permanently, it's just because it's so messy in its current state. I'll see if I'll go on a fixing spree later.

**Comment by morewry** [comment]
In the meantime, a workaround option for using placeholder classes with Compass sprites https://gist.github.com/morewry/4151748

**Comment by scottdavis** [comment]
So are you advocating that the default be placeholders? this is a pretty simple change just need to get the option of those pushing for it.

**Comment by morewry** [comment]
Placeholders as the default? Maybe; don't feel strongly about which the default should be. Opting for placeholders instead of classes? Yes, I'd like to see that option.

**Comment by scottdavis** [comment]
Anyone feel like doing a pull request?

**Comment by aaronjensen** [comment]
Is there a decision on how it should be done? I like the idea of placeholders by default but that would be a breaking change for many. It probably makes more sense to do classes by default for the `@include all-site-sprites;` but placeholders by default for `@import "site/*.png";` since maybe people aren't using the initial class that the import creates other than via mixin?

**Comment by scottdavis** [comment]
Its ok to do a breaking change with a new version release (0.13). So I'm ok setting placeholders to default. The framework functions and mixins would need to be updated along with the content.erb in the sprite importer folder, and all the test files.

**Comment by lucasconstantino** [comment]
Any news on this?

**Comment by morewry** [comment]
https://gist.github.com/morewry/6613650 (different from before) done using `sprite-map` and `sprite-names`. So far as I know still no config option that will change behavior of `@include all-site-sprites`, but perhaps someone else will know differently.

**Comment by scottdavis** [comment]
Probably get to it this weekend I've set aside some time to work on compass

**Comment by chriseppstein** [comment]
> Its ok to do a breaking change with a new version release (0.13).

Nope. not ok. We can deprecate in 0.13 and break in the next release. It's never ok to break an API without first deprecating the usage that will break.

**Comment by chriseppstein** [comment]
As I mentioned in #1448:

Instead of adding a whole new option, I think it would be better to respect the `$<folder>-prefix` variable and have it passed to the `$prefix` argument of the mixins. And then to have the handling of that variable change to allow the inclusion of either a class, placeholder, or id prefix; when none is provided just default to a class selector.

I think this allows people more flexibility in the name of the prefix, keeps the API stable and allows for ID selectors too.

Example:

``` scss
$icon-2x-prefix: '%retina-icon';
@import "icon-2x/*.png";
@include all-icon-2x-sprites;

// $icon-2x-sprite-base-class == "%retina-icon-sprite"
// creates %retina-icon-save, %retina-icon-undo, etc."
```

``` scss
$icon-2x-prefix: '#retina-icon';
@import "icon-2x/*.png";
@include all-icon-2x-sprites;

// Because an ID base class doesn't make sense, default to placeholder for that situation.
// $icon-2x-sprite-base-class == "%retina-icon-sprite"
// creates #retina-icon-save, #retina-icon-undo, etc."
```

**Comment by lucasconstantino** [comment]
Hello, Chris!

So, I just studied the solution you proposed and found some issues. 

**First**, I might be skipping something here but it seams that the `$<folder>-prefix` variable is not currently in use at any place. It's default is `''`, and the place where it should be used (the mixin declarations) instead uses a call to the `sprite-map-name($<folder>-sprites)` function, that basically returns the _folder_ name.
I'm solving this issue right now, modifying both the mixin and variable declaration, for the mixin to make use of the `$<folder>-prefix` value as the default and the second to actually be the value of `<%= name %>`, as there is no sense to use the function here because this would return that same value already hold by the _name_ ruby variable. Am I right?

These changes will allow users to overwrite the prefix variable or use the corresponding mixin argument to make use of the _prefix_ functionality.

**Second**, the `$prefix` argument on the mixins refeers only to the starting name part of the sprite's class selectors: changing it would't change the first character `.` as this is hard coded on the `sprites` mixin. I could, indeed, make this variable be responsible for this base selector string (what I'm calling the dot character, though a better name would maybe "selector kind"), but that **would crash any current use of the mixin argument `$prefix`**, as they would have to add the `.` to their prefix names. As you stated in your comment, that kind of thing shouldn't happen.

**Third**, you give understanding that changing the `$<folder>-prefix` variable should also modify the `$<folder>-sprite-base-class` value, but today this second variable defaults to `.<%= name %>#{$<%= name %>-class-separator}sprite`, that does not make use of the prefix variable. To make it work, we would have to replace the first `<%= name %>` with `#{$<%= name %>-prefix}`. That would work (and wouldn't break any current usage, as for today the prefix variable is simply ignored and that it would now default to the same value as _name_), but we would still have the hard coded `.` at the beginning of the value. If we define the prefix variable to contain the `.` by default (something like `$<%= name %>-prefix : ".<%= name %>" !default;`), that would break that call to `sprite_has_valid_selector` on the sprite mixin, as this functions is not expecting the `.`string.

**Fourth**, although I think that using an already existent variable would be the most desirable choice, I also believe that we should have separation of concerns. What I mean is that the prefix as used today means specifically to name your outputting class names, and should be separated from the functionality of transforming this class selectors into something else; be it `.`, `%`, `#` or even an ` `(empty string).
This decoupled solution would also allow for the creation of a global sprite variable - as is the case of the`$default-sprite-separator`. That would allow us to work with multiple sprites this way (take the variables names as a suggestion only):

``` scss
$sprites-base-selector: '%'; // or any character.

@import "icons/*.png";
@import "icons-2x/*.png";
@import "icons-4x/*.png";

@include all-icons-sprites;
@include all-icons-2x-sprites;
@include all-icons-4x-sprites;
```

instead of having to make a variable definition for each sprite, as follows:

``` scss
$icons-prefix: "%icons";
$icons-2x-prefix: "%icons-2x";
$icons-4x-prefix: "%icons-4x";

@import "icons/*.png";
@import "icons-2x/*.png";
@import "icons-4x/*.png";

@include all-icons-sprites;
@include all-icons-2x-sprites;
@include all-icons-4x-sprites;
```

That only variable would make everything work, while it's absence would keep things working just as before.

---

My point is we have two possible solution
1. Break the current usage of the prefix arguments, making it handle the base selector (be it `.` or whatever) and making the `$<folder>-prefix` variable really be used. Also modify the `sprite_has_valid_selector` function to handle the first character;
2. Introduce a new variable(s).

I would sure go with number two. I do agree with you, thought, that "placeholder" name variables would be to specific. I liked the idea of "any kind selector", as people would be able to use class, IDs, placeholder or even keep it empty so to make usage of sprite for real custom tags selectors.

So, I'm currently working on a new pull request that would follow this premises by doing the following steps:
1. Introduce a `$<folder>-sprite-base-selector` variable that would default to `.`. Changing this would change both `$<folder>-sprite-base-class` and each sprite selector added via the `all-<folder>-sprites` mixin. Note that replacing the `$<folder>-sprite-base-class` would still overwrite this, making it possible for the user to have specific image selectors differs from the "sprite base class" (and now we see this variable ain't semantic anymore);
2. Update all sprites mixins to make use of the introduced variable while keeping things working for old codes;
3. Optionally introduce a `$sprites-base-selectors` global variable, that would be the default value of the `$<folder>-sprite-base-selector`, defaulting itself to `.`. That would be awesome, so the we could have the hole functionality working for different sprite maps at the add of one line of code for the end user.

I'm going to wait for some feedback before making the actual pull request so to avoid an overload of related requests.

---

I'm sorry for this big answer, but this is a big topic for me as I'm working on a new Compass based Sass framework.

**Comment by lucasconstantino** [comment]
@chriseppstein , any thoughts on my previous awnser?
